### PR TITLE
Adds UT for ReconcileLoadBalancers function

### PR DESCRIPTION
**Pull request by carmal891**
**What this PR does / why we need it**: Adds UT for ReconcileLoadBalancers function

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes # https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/issues/1818

**Special notes for your reviewer**:

/area provider/ibmcloud

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Release note**:

```release-note

```

### CODE REVIEW

File: cloud/scope/powervs_cluster_test.go [hunk 1]
```diff
@@ -23,11 +23,13 @@ import (
 
 	"github.com/IBM/go-sdk-core/v5/core"
 	"github.com/IBM/platform-services-go-sdk/resourcecontrollerv2"
+	"github.com/IBM/vpc-go-sdk/vpcv1"
 	"go.uber.org/mock/gomock"
 
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/utils/ptr"
 
+	"sigs.k8s.io/cluster-api-provider-ibmcloud/api/v1beta2"
```
Comment (Karthik-K-N) [inline cloud/scope/powervs_cluster_test.go:32]: I think you should use the existing import `infrav1beta2 "sigs.k8s.io/cluster-api-provider-ibmcloud/api/v1beta2" `

Comment (carmal891) [reply cloud/scope/powervs_cluster_test.go]: Used existing import

**Comment by carmal891** [comment]
@Amulyam24 Adjusted UT as per issue #1937

**Review by Amulyam24 — COMMENTED** [review]
Thank you for the PR @carmal891! 
Minor suggestions, overall LGTM

File: cloud/scope/powervs_cluster_test.go [hunk 2]
```diff
@@ -1093,6 +1095,387 @@ func TestGetResourceGroupID(t *testing.T) {
...
+			Name:               core.StringPtr("test-lb"),
+		}, nil, nil)
+
+		requeue, err := clusterScope.ReconcileLoadBalancers()
+		g.Expect(requeue).To(BeTrue())
+		g.Expect(err).To(BeNil())
+
+		loadBalancerStatus, ok := clusterScope.IBMPowerVSCluster.Status.LoadBalancers["test-lb"]
+		g.Expect(ok).To(BeTrue())
+		g.Expect(loadBalancerStatus.State).To(BeEquivalentTo(infrav1beta2.VPCLoadBalancerStateActive))
```
[the remark this answers is not in the document]
Comment (carmal891) [reply cloud/scope/powervs_cluster_test.go:1207]: done

File: cloud/scope/powervs_cluster_test.go [hunk 3]
```diff
@@ -1093,6 +1095,387 @@ func TestGetResourceGroupID(t *testing.T) {
...
+			ProvisioningStatus: core.StringPtr("active"),
+		}, nil)
+
+		requeue, err := clusterScope.ReconcileLoadBalancers()
+		g.Expect(requeue).To(BeTrue())
+		g.Expect(err).To(BeNil())
+
+		loadBalancer, ok := clusterScope.IBMPowerVSCluster.Status.LoadBalancers["test-lb"]
+		g.Expect(ok).To(BeTrue())
+		g.Expect(loadBalancer.State).To(BeEquivalentTo(infrav1beta2.VPCLoadBalancerStateActive))
```
[the remark this answers is not in the document]
Comment (carmal891) [reply cloud/scope/powervs_cluster_test.go]: Done

File: cloud/scope/powervs_cluster_test.go [hunk 4]
```diff
@@ -1093,6 +1095,387 @@ func TestGetResourceGroupID(t *testing.T) {
...
+					ResourceGroup: &infrav1beta2.IBMPowerVSResourceReference{
+						ID: core.StringPtr("test-resource-gid"),
+					},
+					LoadBalancers: []infrav1beta2.VPCLoadBalancerSpec{
+						{
+							Name: "test-lb",
+							ID:   nil,
+						},
+					},
+					VPCSubnets: nil,
```
[the remark this answers is not in the document]
Comment (carmal891) [reply cloud/scope/powervs_cluster_test.go:1393]: True that the default value for VPCSubnets would be nil, I think explicitly setting it to nil improves readability and clarity. Let me know your opinion

[the remark this answers is not in the document]
Comment (Amulyam24) [reply cloud/scope/powervs_cluster_test.go:1393]: Makes sense 👍🏼

File: cloud/scope/powervs_cluster_test.go [hunk 5]
```diff
@@ -1093,6 +1095,395 @@ func TestGetResourceGroupID(t *testing.T) {
...
+			ProvisioningStatus: core.StringPtr("active"),
+		}, nil, nil)
+
+		requeue, err := clusterScope.ReconcileLoadBalancers()
+		g.Expect(requeue).To(BeFalse())
+		g.Expect(err).To(BeNil())
+
+		loadBalancer, ok := clusterScope.IBMPowerVSCluster.Status.LoadBalancers["test-lb"]
+		g.Expect(ok).To(BeTrue())
+		g.Expect(loadBalancer.State).To(BeEquivalentTo(infrav1beta2.VPCLoadBalancerStateActive))
```
Comment (Amulyam24) [inline cloud/scope/powervs_cluster_test.go:1483]: How about we verify the values of `ControllerCreated` and `Hostname` in status when the resource is created via controller?

**Review by Amulyam24 — COMMENTED** [review]
Overall LGTM apart from a minor comment!

File: cloud/scope/powervs_cluster_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (carmal891) [reply cloud/scope/powervs_cluster_test.go]: corrected as to stay uniform with existing style in codebase

File: cloud/scope/powervs_cluster_test.go (hunk 5 above)
Comment (carmal891) [reply cloud/scope/powervs_cluster_test.go:1483]: Done

File: cloud/scope/powervs_cluster_test.go [hunk 6]
```diff
@@ -1093,6 +1095,395 @@ func TestGetResourceGroupID(t *testing.T) {
...
+				Spec: capiv1beta1.ClusterSpec{
+					ClusterNetwork: &capiv1beta1.ClusterNetwork{
+						APIServerPort: &clusterAPIServerPort,
+					},
+				},
+			},
+		}
+
+		mockVpc.EXPECT().GetLoadBalancerByName(gomock.Any()).Return(nil, nil)
+		mockVpc.EXPECT().CreateLoadBalancer(gomock.Any()).Return(nil, nil, errors.New("Failed loadBalancer creation"))
```
[the remark this answers is not in the document]
Comment (carmal891) [reply cloud/scope/powervs_cluster_test.go]: Done

File: cloud/scope/powervs_cluster_test.go [hunk 7]
```diff
@@ -1093,6 +1095,397 @@ func TestGetResourceGroupID(t *testing.T) {
...
+		setup(t)
+		t.Cleanup(teardown)
+
+		clusterScope := PowerVSClusterScope{
+			IBMVPCClient: mockVpc,
+			IBMPowerVSCluster: &infrav1beta2.IBMPowerVSCluster{
+				Spec: infrav1beta2.IBMPowerVSClusterSpec{
+					LoadBalancers: []infrav1beta2.VPCLoadBalancerSpec{
+						{
+							ID: core.StringPtr("test-lb-instanceid"),
```
Comment (Karthik-K-N) [inline cloud/scope/powervs_cluster_test.go:1125]: Previously we were using core package for pointer conversion, Lately we are using `"k8s.io/utils/ptr"` so lets use ptr.To()
(Wherever applicable in the file)

**Review by Karthik-K-N — COMMENTED** [review]
One small change, Otherwise LGTM

**Comment by Karthik-K-N** [comment]
Just a quick update, @dharaneeshvrd is submitting a fix for https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/issues/2026, So which might require slight change in this UT, so lets keep an eye on this and update it accordingly.

**Comment by carmal891** [comment]
@Karthik-K-N Adjusted as per fix for https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/issues/2026. Additionally added UT for individual sub functions as discussed

File: cloud/scope/powervs_cluster_test.go [hunk 8]
```diff
@@ -1156,6 +1156,851 @@ func TestGetResourceGroupID(t *testing.T) {
...
+							ID: ptr.To("test-lb-instanceid"),
+						},
+					},
+				},
+			},
+		}
+
+		mockVpc.EXPECT().GetLoadBalancer(gomock.Any()).Return(nil, nil, errors.New("failed to fetch VPC load balancer details"))
+
+		requeue, err := clusterScope.ReconcileLoadBalancers()
```
Comment (Amulyam24) [inline cloud/scope/powervs_cluster_test.go:1195]:
```suggestion
		loadBalancerReady, err := clusterScope.ReconcileLoadBalancers()
```
Please change variable and test case name wherever applicable.

File: cloud/scope/powervs_cluster_test.go [hunk 9]
```diff
@@ -1156,6 +1156,851 @@ func TestGetResourceGroupID(t *testing.T) {
...
+		}
+
+		mockVpc.EXPECT().GetLoadBalancerByName(gomock.Any()).Return(nil, nil)
+
+		requeue, err := clusterScope.ReconcileLoadBalancers()
+		g.Expect(requeue).To(BeFalse())
+		g.Expect(err).ToNot(BeNil())
+	})
+
+	t.Run("When createLoadBalancer returns error as resource group id is empty", func(*testing.T) {
```
Comment (Amulyam24) [inline cloud/scope/powervs_cluster_test.go:1429]: I see a couple of redundant test cases here and in `TestCreateLoadbalancer`, can we remove from one function?

File: cloud/scope/powervs_cluster_test.go (hunk 8 above)
Comment (carmal891) [reply cloud/scope/powervs_cluster_test.go]: Corrected

File: cloud/scope/powervs_cluster_test.go (hunk 9 above)
Comment (carmal891) [reply cloud/scope/powervs_cluster_test.go]: Removed redundant cases

**Review by Karthik-K-N — COMMENTED** [review]
/lgtm

**Review by Amulyam24 — APPROVED** [review]

**Review by Karthik-K-N — COMMENTED** [review]
/lgtm

**Review by Prajyot-Parab — APPROVED** [review]
/lgtm

**Pull request merged**